Changes search index format to fasten `pod search --full` command.

manuyavuz opened the pull request:
Detailed discussion can be found here: CocoaPods/cocoapods-search#8 
- Previous implementation was having the drawback of traversing and performing `gsub` on all index strings for each pod specification.
- New implementation stores words as keys and list of pods containing corresponding word inside their specification as the values for corresponding hash keys. Therefore, while searching for a query, we only need to check if query is matched with any key in index hash, if so, we will add corresponding list of spec names to a Set object. Resulted Set object gives us the search result. Using this policy, sources manager can perform a faster search operation.

floere commented:
This is great! :rocket: 

Were you able to run the specs? There are two failures, reproduced here:

```
Pod::Informative: [!] Unable to find a pod with name, author, summary, or descriptionmatching `Chunky`
	/Users/travis/build/CocoaPods/CocoaPods/lib/cocoapods/sources_manager.rb:130:in `search_by_name': In general - can perform a full text search of the sets
	spec/unit/sources_manager_spec.rb:75:in `block (3 levels) in <module:Pod>'
	/Users/travis/build/CocoaPods/CocoaPods/spec/spec_helper/pre_flight.rb:33:in `call'
	/Users/travis/build/CocoaPods/CocoaPods/spec/spec_helper/pre_flight.rb:33:in `block in <class:Context>'
	spec/unit/sources_manager_spec.rb:73:in `block (2 levels) in <module:Pod>'
	spec/unit/sources_manager_spec.rb:37:in `block in <module:Pod>'
	spec/unit/sources_manager_spec.rb:30:in `<module:Pod>'
	spec/unit/sources_manager_spec.rb:29:in `<top (required)>'
Pod::Informative: [!] Unable to find a pod with name, author, summary, or descriptionmatching `Ch[aeiou]nky`
	/Users/travis/build/CocoaPods/CocoaPods/lib/cocoapods/sources_manager.rb:130:in `search_by_name': In general - can perform a full text regexp search of the sets
	spec/unit/sources_manager_spec.rb:82:in `block (3 levels) in <module:Pod>'
	/Users/travis/build/CocoaPods/CocoaPods/spec/spec_helper/pre_flight.rb:33:in `call'
	/Users/travis/build/CocoaPods/CocoaPods/spec/spec_helper/pre_flight.rb:33:in `block in <class:Context>'
	spec/unit/sources_manager_spec.rb:80:in `block (2 levels) in <module:Pod>'
	spec/unit/sources_manager_spec.rb:37:in `block in <module:Pod>'
	spec/unit/sources_manager_spec.rb:30:in `<module:Pod>'
	spec/unit/sources_manager_spec.rb:29:in `<top (required)>'
```

manuyavuz commented:
Yeah, I observed them also, but couldn't have time to fix it. Hoping to fix these tomorrow.

floere commented:
That's great! :rocket:

manuyavuz commented:
Hey @floere,

I checked again and discovered that failures were caused by Core repo here not containing my changes in CocoaPods/Core#265.

How should I manage this? This PR is tightly coupled to changes in Core repo, so I guess it will not pass unless the PR in Core repo is merged.

Need suggestions.

floere commented:
Are the Core changes standalone? (Thery work independently?)

On Sun, Oct 4, 2015 at 3:09 PM, Muhammed Yavuz Nuzumlali
<notifications@github.com> wrote:

> Hey @floere,
> I checked again and discovered that failures were caused by Core repo here not containing my changes in CocoaPods/Core#265.
> How should I manage this? This PR is tightly coupled to changes in Core repo, so I guess it will not pass unless the PR in Core repo is merged.
> Need suggestions.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/CocoaPods/CocoaPods/pull/4249#issuecomment-145349241

manuyavuz commented:
Core include changes on `aggregate.rb` and can run standalone. However, `sources_manager.rb` here is dependent on the changes in aggregate code. Therefore, CocoaPods/Core#265 can pass individually, but this won't pass.

segiddins commented:
You can point the gem file here to your branch until the Core PR is merged

manuyavuz commented:
@segiddins I tried your suggestion but probably missing something.

Added the following line to Gemfile of CocoaPods project:

```
  gem 'cocoapods-core', :git => 'https://github.com/manuyavuz/Core.git', :branch => 'improvement/searchPerformance'
```

When I run rake spec I get the following error:

```
[!] There was an error parsing `Gemfile`: You cannot specify the same gem twice coming from different sources.
You specified that cocoapods-core (>= 0) should come from https://github.com/CocoaPods/Core.git (at master) and https://github.com/manuyavuz/Core.git (at improvement/searchPerformance)
. Bundler cannot continue.

 #  from /Users/manuyavuz/Development/OpenSource/Cocoapods/Rainforest/CocoaPods/Gemfile:29
 #  -------------------------------------------
 #    gem 'cocoapods-dependencies'
 >    gem 'cocoapods-core', :git => 'https://github.com/manuyavuz/Core.git', :branch => 'improvement/searchPerformance'
 #    gem 'bacon'
 #  -------------------------------------------
```

Is this default Core dependency come from `gem 'cocoapods-dependencies'`? If so, how can I solve this? Setting local dependency won't work because travis do not run using local dependencies I guess?

manuyavuz commented:
Ah, got it. Removed cp_gem line for Core dependency.

manuyavuz commented:
@floere, previous errors have gone now. However, there are two fails for functional specs. I was not observing them in the previous commit. 

I think they may be related to changes included from that time until now because I rebased the branch with master before sending the last commit. 

Also the failure seems not related to any source file that I've changed.

manuyavuz commented:
Hi, 

I've completed update index policy. All tests pass but got a strange error during running Inch.

```
--------------------------------------------------------------------------------
Running Inch
--------------------------------------------------------------------------------
Parse docs …
.............................................................................................
Evaluating …
................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
rake aborted!
ArgumentError: wrong number of arguments (2 for 1)
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/colored-1.2/lib/colored.rb:83:in `color'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/term-ansicolor-1.3.0/lib/term/ansicolor.rb:226:in `on_color'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/utils/ui.rb:65:in `__header'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/utils/ui.rb:47:in `header'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/cli/command/output/list.rb:27:in `block in display_list'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/cli/command/output/list.rb:22:in `each'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/cli/command/output/list.rb:22:in `display_list'
/Users/travis/build/CocoaPods/CocoaPods/vendor/bundle/ruby/2.0.0/gems/inch-0.5.10/lib/inch/cli/command/output/list.rb:16:in `initialize'
/Users/travis/build/CocoaPods/CocoaPods/Rakefile:344:in `new'
/Users/travis/build/CocoaPods/CocoaPods/Rakefile:344:in `block (2 levels) in <top (required)>'
/Users/travis/build/CocoaPods/CocoaPods/Rakefile:146:in `block (2 levels) in <top (required)>'
Tasks: TOP => inch:spec
(See full trace by running task with --trace)
```

What could be the reason?

segiddins commented:
That happens when inch fails, so something isn't properly documented.

manuyavuz commented:
@floere with the last commit all tests pass. You may check out them now.

segiddins commented:
W00t! Excellent work, @manuyavuz , really look forward to reviewing this! If you want, feel free to give me your email so we can get you into our slack org?

manuyavuz commented:
Thanks @segiddins! I would be very happy to join the slack team because I'm hoping to contribute more. You can use the email in my github profile.

Waiting to hear for more!

segiddins commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -208,11 +253,14 @@ def update(source_name = nil, show_output = false)
           UI.section "Updating spec repo `#{source.name}`" do
             Dir.chdir(source.repo) do
               begin
+                prev_commit_hash = (git! %w(rev-parse HEAD)).strip
```
can we move this updating logic into `Source` ? i.e. so it has an `update` method that returns the changes spec paths?

manuyavuz replied:
You mean moving also the following, right?

```
output = git! %w(pull --ff-only)
```

segiddins replied:
yup!

manuyavuz replied:
@segiddins should we also move UI code there into source?

I don't think it's suitable because we would need config object inside source to handle UI things.

I'm thinking of making a method returning a tuple variable including git pull output and changed paths. It would be sth like this:

```
[output, changed_spec_paths]  = source.update
```
Does this fit?

segiddins commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -224,6 +272,11 @@ def update(source_name = nil, show_output = false)
           end
         end
+        # Perform search index update operation as a subprocess.
+        fork do
```
should the forked process do `Process.daemon` so that it doesn't keep the parent process alive?

segiddins commented on spec/unit/sources_manager_spec.rb:
```diff
@@ -241,11 +231,12 @@ module Pod
-        SourcesManager.expects(:git!).with { |options| options.should.include? '--ff-only' }
-        SourcesManager.update(test_repo_path.basename.to_s, true)
-      end
+      # TODO: how should we change this? Update code runs `git rev-parse HEAD` which fails for this spec
```
stub out the git things to return what you need them to return

segiddins commented:
Yup

segiddins commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -142,41 +137,64 @@ def search_by_name(query, full_text_search = false)
+      def save_search_index(index)
+        require 'json'
+        @updated_search_index = index
+        File.open(search_index_path, 'w') do |file|
```
`search_index_path` should be a `Pathname`

manuyavuz replied:
I think this way works too. Current implementation also performs open in the same way as the following:

```
      File.open(search_index_path, 'w') do |file|
        file.write(search_index.to_yaml)
      end
```

What kind of problem may we face?

segiddins replied:
No problem, we just try to stay away from using strings as paths and use pathnames instead

manuyavuz replied:
I think it indeed gives a pathname. It's a method of sources_manager as the following (line 202):

```
      # @return [Pathname] The path where the search index should be stored.
      #
      def search_index_path
        Config.instance.search_index_file
      end
```

segiddins replied:
in that case, lets just use `search_index_path.open`

segiddins commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -208,22 +255,19 @@ def update(source_name = nil, show_output = false)
           end
         end
+        # Perform search index update operation as a subprocess.
+        Process.fork do
```
can we extract this bit into a new method?
`update_search_index_if_needed_in_background` or something?

segiddins commented on spec/unit/sources_manager_spec.rb:
```diff
@@ -241,11 +231,12 @@ module Pod
-        SourcesManager.expects(:git!).with { |options| options.should.include? '--ff-only' }
-        SourcesManager.update(test_repo_path.basename.to_s, true)
-      end
+      # TODO: given that update moved to source.rb, should this still exist here?
```
this is good to delete

manuyavuz commented:
I've completed implementation and fixed multi word search issue. Solving that required a change in cocoapods-search plugin also (CocoaPods/cocoapods-search#10)

I think this is ready for final review too.

floere commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -192,6 +214,44 @@ def search_index_path
+        changed_spec_paths.each_pair do |source, spec_paths|
+          source_name = source.name
+          next unless spec_paths.length > 0
+          search_index = stored_search_index
```
Could this be `||=`?

manuyavuz replied:
We do not want to create a search index from scratch while updating repos if there is no search index created before. So i think it should stay as it is, and exit from block at line 229.

floere replied:
What do you mean when you say "exit from block"?

manuyavuz replied:
Ah man I got the possible bug now. You're right. Fixing it.

floere commented on lib/cocoapods/sources_manager.rb:
```diff
@@ -192,6 +214,44 @@ def search_index_path
+          # First traverse search_index and update existing words
+          # Removed traversed words from new_index after adding to search_index,
+          # so that only non existing words will remain in new_index after enumeration completes.
+          search_index[source_name].each_pair do |word, _|
```
You could store the result of the `search_index[source_name]` access in `source_search_index` (example name) to clarify that it's always about the same piece of the search index.

floere commented:
Only have two small comments. Thanks @manuyavuz!

manuyavuz commented:
Consider this case:

- User performs a repo update.
- Update method detects changes in one of the repositories.
- There is an existing search_index, but it does not contain inverted index for updated repo (possibly because user added repo after she performed a pod search).

Should this code generate an inverted index for only the changed files? Or should it ignore them?

manuyavuz commented:
This decision will also affect search policy.

I think the most appropriate flow is the following:

- During repo update, search index for a source repo is updated only if there exist a previous search index generated for the source repo. If there is no index for the source, it will not be generated, even if there are changes in the repo.

- During a `pod search --full` search, code will look for each source, and will generate search index from scratch if there is no index generated for that source before.

If that's OK, i'll finish development accordingly.

Thoughts, @floere, @segiddins, @alloy?

floere commented:
@manuyavuz That makes a lot of sense to me.

manuyavuz commented:
Guys, I have a problem. 

As you can see, travis fails on ` updates a repository` spec of `update_spec.rb` file.

Because the new update method starts a daemon process in background, it keeps search_index.json file open. Specs try to clear `tmp` folder in `preflight.rb` file before any specs, but because the file is in use by daemon process, remove fails with `Errno::ENOTEMPTY` error.

I've solved this in unit test via stubbing `update_search_index_if_needed_in_background` method which starts the daemon process. However, I'm not sure if this is the correct way to handle it in a functional spec. 

What should I do to resolve this?

segiddins commented:
@manuyavuz make it so `update_search_index_if_needed_in_background` returns a PID you can wait on?

manuyavuz commented:
@segiddins Wouldn't original update method (therefore the `pod repo update` command) be blocked if I wait for that pid to finish?

segiddins commented:
Good point. I'd just expect that the background method get's called then, manually call the synchronous method, and then verify

manuyavuz commented:
I actually did as you suggest in unit tests of SourcesManager class. However, not sure if this is an appropriate thing to do in a functional spec. Failed test is the following:

``` ruby
    it 'updates a repository' do
      upstream = SpecHelper.temporary_directory + 'upstream'
      FileUtils.cp_r(test_repo_path, upstream)
      Dir.chdir(test_repo_path) do
        `git remote add origin #{upstream}`
        `git remote -v`
        `git fetch -q`
        `git branch --set-upstream-to=origin/master master`
      end
      lambda { command('repo', 'update').run }.should.not.raise
    end
```

Is it alright to change it like the following?:

``` ruby
    it 'updates a repository' do
      upstream = SpecHelper.temporary_directory + 'upstream'
      FileUtils.cp_r(test_repo_path, upstream)
      Dir.chdir(test_repo_path) do
        `git remote add origin #{upstream}`
        `git remote -v`
        `git fetch -q`
        `git branch --set-upstream-to=origin/master master`
      end
      # Add expect for bg method
      SourcesManager.expects(:update_search_index_if_needed_in_background).with({}).returns(nil)
      lambda { command('repo', 'update').run }.should.not.raise
    end
```

If this is OK, I'll make a last rebase and push final version.

segiddins commented:
IMHO that's fine

manuyavuz commented:
Offenses made me sick 😤

I think this is ready to 🚀 also. You can now make a last glance.

segiddins commented on spec/unit/sources_manager_spec.rb:
```diff
@@ -235,28 +225,47 @@ module Pod
         UI.output.should.match /is up to date/
       end
 
-      it 'uses the only fast forward git option' do
```
This test can come back

segiddins commented:
😄 super well done!

floere commented:
:+1:

manuyavuz commented:
It appeared again 😉

segiddins commented:
:rocket: :ship:
Great work @manuyavuz!

manuyavuz commented:
Great!
